### PR TITLE
[RDY] Add neovim memory tmps support, also add PDumpfile to debug.

### PR DESCRIPTION
**Pull request by syswow**
Make vim-clang in neovim faster.

### CODE REVIEW

**Comment by syswow** [comment]
Any advice is welcome. :smiling_imp:

**Comment by justmao945** [comment]
Thank you for your hard work! Seems changed a lot of code, so I need time review~ :smile:

File: plugin/clang.vim [hunk 1]
```diff
@@ -152,6 +152,22 @@ func! s:PLog(head, info)
...
+  let l:is_windows = has('win32') || has('win64')
+  if l:lv <= g:clang_debug && !l:is_windows
+    let l:fn = a:filename
+    if l:fn == ''
+      let l:fn = 'vim-clang-debug.log'
+    endif
+    let l:fullpath = '/tmp/'.l:fn
+    call writefile(a:data, l:fullpath)
+  endif
+endf
```
Comment (justmao945) [inline plugin/clang.vim]: I don't think this function necessary, although output of clang may be large...

File: plugin/clang.vim [hunk 2]
```diff
@@ -841,13 +857,57 @@ func! s:ClangCompleteInit(force)
...
 "{{{ ClangExecuteNeoJobHandler
-"handles event: exit
+"Handles events: stdout, stderr, exit
+"Tmps in memory
 func! ClangExecuteNeoJobHandler(job_id, data, event)
-  if a:event == 'exit'
-    call ClangExecuteDone(self.fstdout, self.fstderr)
+  if a:event == 'stdout'
+    call s:PDebug("ClangExecuteNeoJobHandler", "stdout", 2)
+    if len(b:clang_memtmps[0]) != 0
```
Comment (justmao945) [inline plugin/clang.vim]: I think you can use the same buffer var `b:clang_state`, you can set `stdout` and `stderr` in this dict.

File: plugin/clang.vim [hunk 3]
```diff
@@ -875,9 +935,20 @@ func! s:ClangExecute(root, clang_options, line, col)
...
+    call add(l:nvimsrc, "\<cr>")
+
+    let l:nvimcmd = printf('%s -cc1 -x c++ -fsyntax-only -code-completion-macros -code-completion-at -:%d:%d %s',
+                      \ g:clang_exec, a:line, a:col, a:clang_options)
+    let l:argv = l:nvimcmd
+
+    " set env
+    let l:jobid = jobstart(['sh', '-c', l:argv], s:neojobcallbacks)
+    call jobsend(l:jobid, l:nvimsrc)
+    call jobclose(l:jobid, 'stdin')
```
Comment (justmao945) [inline plugin/clang.vim]: Good job! We can send source content to clang in nvim now!
@mlen here's a solution provided by @syswow !

File: plugin/clang.vim [hunk 4]
```diff
@@ -875,9 +935,20 @@ func! s:ClangExecute(root, clang_options, line, col)
...
+
+    let l:nvimsrc = getline(1, '$')
+    call add(l:nvimsrc, "\<cr>")
+
+    let l:nvimcmd = printf('%s -cc1 -x c++ -fsyntax-only -code-completion-macros -code-completion-at -:%d:%d %s',
+                      \ g:clang_exec, a:line, a:col, a:clang_options)
+    let l:argv = l:nvimcmd
+
+    " set env
+    let l:jobid = jobstart(['sh', '-c', l:argv], s:neojobcallbacks)
```
Comment (justmao945) [inline plugin/clang.vim]: I have tested on my machine, you can just put callbacks in local var, *please* put them in a local dict, `s:neojobcallbacks` is not good.
e.g.
```viml
jobstart(l:argv, {'fstdout': l:tmps[0], 'fstderr': l:tmps[1], 'on_exit': function('ClangExecuteNeoJobHandler')})
```

**Comment by justmao945** [comment]
Very great patch, which will be perfect if you can help to make code cleaner.

File: plugin/clang.vim (hunk 1 above)
Comment (syswow) [inline plugin/clang.vim]: i think it's useful to debug large output, i usually use meld to diff vim-clang output and clang output.

File: plugin/clang.vim (hunk 3 above)
Comment (syswow) [inline plugin/clang.vim]: yes.

File: plugin/clang.vim (hunk 2 above)
Comment (syswow) [inline plugin/clang.vim]: but vim doesn't use it. I think separate it from `b:clang_state` is good.

**Comment by syswow** [comment]
@justmao945  yes, i will make code cleaner later. add some commits for refactor and rebase.

File: plugin/clang.vim (hunk 2 above)
Comment (justmao945) [inline plugin/clang.vim]: What do you mean vim doesn't use it? Everything produced by clang should be put into this buffer var.

File: plugin/clang.vim (hunk 1 above)
Comment (justmao945) [inline plugin/clang.vim]: This function is different from any other Pxxx functions(P here means print). So I don't think is a good idea that we put some debug info into messages and some other into a file, may be confused. I recommend to use `PDebug`

Comment (syswow) [inline plugin/clang.vim]: But if we want to debug `std::` complete, vim-clang will generates tons of lines, and print these lines to screen is meaningless. It's difficult to comfirm whethe these output is correct.

File: plugin/clang.vim (hunk 2 above)
Comment (syswow) [inline plugin/clang.vim]: You are right. In the beginning I think avoiding the mix of nvim buffer with vim buffer is good. Now, I think it is better to unify them.

**Comment by syswow** [comment]
@justmao945  Hi, Sorry for leaving this PR long time on inactive status, because I'm busy recently. So I am going to create a new PR. :smile:

**Comment by justmao945** [comment]
@syswow It's OK. We made this plugin because we all love VIM. Anytime to contribute is accepted :smile:

**Pull request closed without merging**